#!/usr/bin/env python3

import sys
import os
import configparser
import logging
import logging.handlers

from time import sleep

import requests

import dns.resolver


logging.basicConfig(filename='/tmp/certbot_validator.log', level=logging.DEBUG)
logger = logging.getLogger("certbot_validator")
logger.addHandler(logging.StreamHandler(sys.stdout))

class GoDaddySSOAuth(requests.auth.AuthBase):
      def __init__(self, api_key, api_secret):
          self.api_key = api_key
          self.api_secret = api_secret
      def __call__(self, r):
          r.headers['Authorization'] = f"sso-key {self.api_key}:{self.api_secret}"
          return r

def setup():
    logger.info("---Begin Certbot validation script---")

    config = configparser.ConfigParser()
    godaddy_api_key = ""
    godaddy_api_secret = ""
    config_file = os.path.dirname(os.path.abspath(__file__)) + '/godaddy_config.ini'
    if os.path.isfile(config_file):
        logger.info(f"Using {config_file} as config file for api key/secret.")
        failed_config_read = False
        config = configparser.ConfigParser()
        try:
            config.read(config_file)
        except configparser.MissingSectionHeaderError:
            logger.error("godaddy_config.ini not a valid config file. Missing section header(s).")
            finish(False) 
        if 'default' not in config.sections():
            logger.error("Error reading config file, 'default' section not present.")
            finish(False)
        if 'api_key' not in config['default'].keys():
            logger.error("Error reading config file, 'api_key' not found in 'default' section.")
            failed_config_read = True

        if 'api_secret' not in config['default'].keys():
            logger.error("Error reading config file, 'api_secret' not found in 'default' section")
            failed_config_read = True

        if failed_config_read:
            finish(False)
        godaddy_api_key = config['default']['api_key']
        godaddy_api_secret = config['default']['api_secret']
    else:
        logger.info("Using env vars for api key/secret.")
        failed_env_read = False
        # Look in env vars for GODADDY_API_KEY and GODADDY_API_SECRET
        godaddy_api_key = os.getenv('GODADDY_API_KEY')
        godaddy_api_secret = os.getenv('GODADDY_API_SECRET')
        if godaddy_api_key == None:
            logger.error("GODADDY_API_KEY envrionment variable not set.")
            failed_env_read = True
        if godaddy_api_key == None:
            logger.error("GODADDY_API_SECRET envrionment variable not set.")
            failed_env_read = True
        if failed_env_read:
            finish(False)

    #Get the domain from the CERTBOT_DOMAIN and extract the root domain.
    certbot_domain = os.getenv("CERTBOT_DOMAIN")
    if certbot_domain == None:
        logger.error("CERTBOT_DOMAIN env is not set. If you are not running this script for testing, please set this envionment variable")
        finish(False)
    root_domain = f"{certbot_domain.split('.')[-2]}.{certbot_domain.split('.')[-1]}"

    # Test auth and ensure it is correct
    test_godaddy_auth(godaddy_api_key, godaddy_api_secret, root_domain)

    add_godaddy_record(godaddy_api_key, godaddy_api_secret, certbot_domain, root_domain)

def test_godaddy_auth(godaddy_api_key, godaddy_api_secret, root_domain):
    test_url = f"https://api.godaddy.com/v1/domains/{root_domain}/records"
    response = requests.get( test_url, auth=GoDaddySSOAuth(godaddy_api_key, godaddy_api_secret))
    if response.status_code != 200:
        logger.error(f"Auth against go daddy api failed with {response.status_code}.")
        finish(False)

def add_godaddy_record(godaddy_api_key, godaddy_api_secret, certbot_domain, root_domain):
    record_name = "_acme-challenge"

    if certbot_domain != root_domain:
        record_name = f"{record_name}.{certbot_domain.replace('.'+root_domain, '')}"
    
    # Get the validation token:
    validation_token = os.getenv("CERTBOT_VALIDATION", "TESTING123")
    
    
    logger.info(f"Adding record of TXT with name {record_name}")

    godaddy_record_url = f"https://api.godaddy.com/v1/domains/{root_domain}/records/TXT/{record_name}"
    data = [{"data": validation_token, "ttl":600},]
    response = requests.put(godaddy_record_url, auth=GoDaddySSOAuth(godaddy_api_key, godaddy_api_secret), json=data)
    if response.content != b'':
        logger.error(f"Request to go daddy failed with {response.content} as the response.")
        finish(False)

    #Check DNS (using google) to ensure propagation of TXT record
    resolver = dns.resolver.Resolver()
    resolver.nameservers = ['8.8.8.8']
    for dns_try in range(1, 11):
        sleep(5)
        try:
            {%- if (grains['os'] == 'Ubuntu' and grains['osmajorrelease'] < 22) or (grains['os_family'] == 'RedHat' and grains['osmajorrelease'] < 9) or (grains['os'] == 'Debian' and grains['osmajorrelease'] < 11) %}
            answer = resolver.query(f"{record_name}.{root_domain}",'TXT').response.answer[0]
            {%- else %}
            answer = resolver.resolve(f"{record_name}.{root_domain}",'TXT').response.answer[0]
            {%- endif %}
            if validation_token in str(answer):
                logger.info("Verification token applied successfully")
                finish()
        except dns.resolver.NXDOMAIN: 
            if dns_try != 10:
                logger.info("Change has not propagated trying again in 10 seconds.")
                sleep(10)
            else:
                logger.error("Change did not propagate in time. Unknown as to why. Maybe try again?")
                finish(False)



        

def finish(success=True):
    logger.info("---Ending Certbot validation script---")
    if success: 
        sys.exit()
    sys.exit(1)

if __name__ == "__main__":
    setup()


